src/ray/raylet/local_object_manager.cc

```diff
@@ -95,18 +96,27 @@ void LocalObjectManager::PinObjectsAndWaitForFree(
 }
 
 void LocalObjectManager::ReleaseFreedObject(const ObjectID &object_id) {
-  RAY_LOG(DEBUG) << "Unpinning object " << object_id;
-  if (!objects_waiting_for_free_.erase(object_id)) {
+  // Only free the object if it is not already freed.
+  auto it = local_objects_.find(object_id);
+  if (it == local_objects_.end() || it->second.second) {
     return;
   }
-  // The object should be in one of these stats. pinned, spilling, or spilled.
+  it->second.second = true;
+
+  RAY_LOG(DEBUG) << "Unpinning object " << object_id;
+  // The object should be in one of these states: pinned, spilling, or spilled.
   RAY_CHECK((pinned_objects_.count(object_id) > 0) ||
             (spilled_objects_url_.count(object_id) > 0) ||
             (objects_pending_spill_.count(object_id) > 0));
-  spilled_object_pending_delete_.push(object_id);
   if (pinned_objects_.count(object_id)) {
     pinned_objects_size_ -= pinned_objects_[object_id]->GetSize();
     pinned_objects_.erase(object_id);
```

Comment:
will there be a race condition that one owner just pinned the object and immediately unpinned by another owner?

Reply:
I think it's possible at the moment because of #20713 but I don't think it will be possible to have different owners for the same object once we merge #21395.

Reply:
Should we comment this block works assuming there's a single owner?

Reply:
Actually I think this block is fine because we detect if the owner has changed. The bug will surface at the application level, not here.